[WebDriver BiDi] Implement script.realmCreated and script.realmDestroyed events

sergedeh opened the pull request:
#### ccec5e43c6991c10e7e5ba4b8924235930e27255
<pre>
[WebDriver BiDi] Implement script.realmCreated and script.realmDestroyed events
<a href="https://bugs.webkit.org/show_bug.cgi?id=304062">https://bugs.webkit.org/show_bug.cgi?id=304062</a>
<a href="https://rdar.apple.com/166875337">rdar://166875337</a>

Reviewed by BJ Burg.

This patch implements the script.realmCreated and script.realmDestroyed events
as specified in the W3C WebDriver BiDi specification. These events notify BiDi
clients when JavaScript execution realms are created and destroyed in browsing
contexts, enabling proper tracking of realm lifecycle for script execution.

* Source/WebCore/automation/AutomationInstrumentation.cpp:
(WebCore::AutomationInstrumentation::scriptRealmCreated): Added with world filtering.
(WebCore::AutomationInstrumentation::scriptRealmDestroyed): Added with world filtering.

* Source/WebCore/automation/AutomationInstrumentation.h:
Added scriptRealmCreated and scriptRealmDestroyed methods with DOMWrapperWorld parameter.

* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::resolveOriginForRealm): Added helper to resolve realm origin.
(WebCore::WindowProxy::destroyJSWindowProxy): Notify realm destruction, with optional frameForNotification parameter.
(WebCore::WindowProxy::createJSWindowProxy): Notify realm creation.
(WebCore::WindowProxy::setDOMWindow): Notify realm destruction and re-creation during navigation.
(WebCore::WindowProxy::detachFromFrame): Save frame reference before nullifying for realm destruction notifications.

* Source/WebCore/bindings/js/WindowProxy.h:
Updated destroyJSWindowProxy signature to accept optional Frame pointer for notifications during detachment.

* Source/WebDriver/Session.cpp:
(WebDriver::extractEventBrowsingContextIDs): Helper to extract context IDs from event parameters.
(WebDriver::haveSameElements): Helper to compare vectors for subscription matching.
(WebDriver::subscriptionMatchesScope): Check if subscription matches event scope.
(WebDriver::shouldReplayRealmCreatedEvents): Check if realm replay is needed.
(WebDriver::removeSubscribedEvents): Helper to decrement event subscription counts.
(WebDriver::Session::dispatchBidiMessage): Add context-based event filtering and realm event replay.
(WebDriver::Session::processLogEntryAdded): Extract event parameters for context filtering.

* Source/WebDriver/Session.h:
Changed emitActiveBidiScriptRealmCreatedEvents to return void.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::emitActiveBidiScriptRealmCreatedEvents): Emit realm events for active realms.
Handle subscribe command with realm event replay support.

* Source/WebKit/Scripts/webkit/messages.py:
Added WebKit::RealmIdentifier to serialized_identifiers() list for proper IPC message generation.

* Source/WebKit/Shared/IdentifierTypes.h:
Added RealmIdentifier as ObjectIdentifier&lt;RealmIdentifierType&gt;.

* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
Added RealmIdentifierType template for ObjectIdentifier serialization.

* Source/WebKit/UIProcess/Automation/Automation.json:
Added emitActiveBidiScriptRealmCreatedEvents command.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::notifyRealmCreated): Track realm and emit script.realmCreated event.
(WebKit::BidiScriptAgent::notifyRealmDestroyed): Remove realm and emit script.realmDestroyed event.
(WebKit::BidiScriptAgent::realmIdentifierForBrowsingContext): Lookup helper for UIProcess-initiated destruction.
(WebKit::BidiScriptAgent::generateRealmIdForBrowsingContext): Generate realm ID string from identifier.
(WebKit::BidiScriptAgent::emitEventsForActiveRealms): Emit realm events for all active realms.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
Added RealmInfo struct for tracking realm metadata.
Added RealmIdentifier-based realm tracking (m_browsingContextToRealmId, m_activeRealms).
Added notifyRealmCreated, notifyRealmDestroyed, realmIdentifierForBrowsingContext, and emitEventsForActiveRealms methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::createBrowsingContext): Call ensureRealmForInitialEmptyDocument.
(WebKit::WebAutomationSession::scriptRealmCreated): Handle realm creation IPC from WebProcess using RealmIdentifier.
(WebKit::WebAutomationSession::scriptRealmDestroyed): Handle realm destruction IPC with activeRealms lookup to avoid race conditions.
(WebKit::WebAutomationSession::contextDestroyedForPage): Lookup realm ID before notifying destruction.
(WebKit::WebAutomationSession::emitActiveBidiScriptRealmCreatedEvents): Delegate to BidiScriptAgent.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Added scriptRealmCreated, scriptRealmDestroyed, and emitActiveBidiScriptRealmCreatedEvents handler declarations using RealmIdentifier type.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.messages.in:
Added ScriptRealmCreated and ScriptRealmDestroyed IPC messages with RealmIdentifier parameter.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor): Initialize m_scriptDomainNotifier.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
Added scriptAgent() and scriptDomainNotifier() accessors.
Added m_scriptDomainNotifier member.

* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
Added realmCreated and realmDestroyed event definitions.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::scriptRealmCreated): Forward to WebAutomationSession via IPC with RealmIdentifier.
(WebKit::WebAutomationSessionProxy::scriptRealmDestroyed): Forward to WebAutomationSession via IPC with RealmIdentifier.
(WebKit::WebAutomationSessionProxy::ensureRealmForInitialEmptyDocument): Force JSWindowProxy creation using jsWindowProxy().

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
Added AutomationInstrumentationClient interface implementation.
Added ensureRealmForInitialEmptyDocument method.
Added m_frameToRealmIdentifier mapping.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
Added EnsureRealmForInitialEmptyDocument IPC message.

* WebDriverTests/TestExpectations.json:
Updated test expectations for realm_created tests to match current implementation.
Added missing test expectations for sandbox, workers, and iframe tests.

Canonical link: <a href="https://commits.webkit.org/310527@main">https://commits.webkit.org/310527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6d05c26407a10a7ad565312815816ada1e16798

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/154097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/27354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/20512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/162849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/107563 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/155970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/27486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/27205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/119177 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/84255 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/157056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/21414 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/138376 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/99877 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/153417 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/20510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/18499 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/10682 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/130172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/16227 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/165322 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/8537 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/17828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/127267 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/26900 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/22537 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/127418 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/34573 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/26823 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/138021 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/83410 "The change is no longer eligible for processing.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/22288 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/14809 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/26515 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/90606 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/26095 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/26325 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/26167 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -107,6 +128,15 @@ void WindowProxy::destroyJSWindowProxy(DOMWrapperWorld& world)
+    // Only notify for normal world (main page execution), not user/internal worlds.
+    if (world.isNormal() && m_frame) {
+        if (RefPtr localFrame = dynamicDowncast<LocalFrame>(*m_frame))
+            AutomationInstrumentation::scriptRealmDestroyed(localFrame->frameID());
```
It would be better to pass the `world` unconditionally to `AutomationInstrumentation` to avoid leaking WebDriver logic into core WebCore classes.

burg commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -122,6 +152,16 @@ JSWindowProxy& WindowProxy::createJSWindowProxy(DOMWrapperWorld& world)
+#if ENABLE(WEBDRIVER_BIDI)
+    // Notify about realm creation for automation.
+    // Only notify for normal world (main page execution), not user/internal worlds.
+    if (world.isNormal()) {
```
Ditto.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -172,8 +172,8 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
 
-    String realmID = generateRealmIdForBrowsingContext(*browsingContext);
-    session->evaluateJavaScriptFunction(topLevelContextHandle, frameHandle, functionDeclaration, WTF::move(argumentsArray), false, optionalUserActivation.value_or(false), std::nullopt, [callback = WTF::move(callback), realmID](Inspector::CommandResult<String>&& stringResult) {
+    String realmId = generateRealmIdForBrowsingContext(*browsingContext);
```
Strongly prefer realmID over realmId. Per WebKit style it should be realmID or realmIdentifier, but not realmId.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -458,74 +456,34 @@ RefPtr<Inspector::Protocol::BidiScript::RealmInfo> BidiScriptAgent::createRealmI
+    }
+
+    // Use the shared m_realmNavigationCounters to ensure consistency with notifyRealmCreated/Destroyed.
+    auto counterIterator = m_realmNavigationCounters.find(*contextHandle);
```
Nit: `it`

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -458,74 +456,34 @@ RefPtr<Inspector::Protocol::BidiScript::RealmInfo> BidiScriptAgent::createRealmI
-    // Update the cache with the new realm ID.
-    m_frameRealmCache.set(frameInfo.frameID, FrameRealmCacheEntry { currentURL, currentDocumentID, newRealmId });
+    uint64_t counter = counterIterator->value;
+    String realmId = counter <= 1
```
I'm not a fan of string identifiers, can we introduce `RealmIdentifier` as an ObjectIdentifier subclass?

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -648,13 +599,86 @@ void BidiScriptAgent::collectExecutionReadyFrameRealms(const FrameTreeNodeData&
+    eventParameters->setString("context"_s, browsingContext);
+    eventObject->setObject("params"_s, WTF::move(eventParameters));
+
+    session->sendBidiMessage(eventObject->toJSONString());
```
This should be using ScriptFrontendDispatcher to do the message building.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -648,13 +599,86 @@ void BidiScriptAgent::collectExecutionReadyFrameRealms(const FrameTreeNodeData&
+        return;
+
+    uint64_t counter = counterIterator->value;
+    String realmId = counter == 1
```
The counter idea is the same thing as IdentifiersFactory. Maybe just use that? Again, i'd prefer we pass around a strongly typed RealmIdentifier that can be used in a WeakHashMap to look up the JSWindowProxy and do stuff with it.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -648,13 +599,86 @@ void BidiScriptAgent::collectExecutionReadyFrameRealms(const FrameTreeNodeData&
+    eventParameters->setString("realm"_s, realmId);
+    eventObject->setObject("params"_s, WTF::move(eventParameters));
+
+    session->sendBidiMessage(eventObject->toJSONString());
```
Ditto.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -69,7 +63,18 @@ class BidiScriptAgent final : public Inspector::BidiScriptBackendDispatcherHandl
+    struct RealmInfo {
+        String realmId;
+        String origin;
+        String type; // "window", "dedicated-worker", etc.
```
I think all of these can be strongly typed.

RealmIdentifier
Origin / Website
RealmType
BrowsingContextHandle

sergedeh replied:
For the Origin field, I picked SecurityOriginData. SecurityOrigin is carrying too much data with it (Ref counting etc). There is also RegistrableDomain (the "website" type), but that only captures the top-level domain like "example.com" - it loses protocol, port, and subdomain info that the BiDi spec needs for realms.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1211,6 +1218,9 @@ void WebAutomationSession::contextDestroyedForPage(const WebPageProxy& page)
     auto [clientWindow, userContext] = getClientWindowAndUserContext(page);
 
+    // Ensure the active realm is destroyed even if the WebProcess terminates first.
+    m_bidiProcessor->scriptAgent().notifyRealmDestroyed(contextHandle);
```
Seems like some of these same parameters as below might be needed?

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -148,6 +148,10 @@
 #include "WKContextPrivate.h"
 #include "WebAutomationSession.h"
 #include "WebAutomationSessionProxyMessages.h"
+#if ENABLE(WEBDRIVER_BIDI)
```
Pls put this block of includes below, where other conditional includes are located.

burg commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -160,6 +160,12 @@ messages -> WebPageProxy {
+
+#if ENABLE(WEBDRIVER_BIDI)
+    # WebDriver BiDi realm lifecycle events
+    ScriptRealmWasCreated(WebCore::FrameIdentifier frameID, String origin) -> ()
```
Shouldn't this pass the RealmIdentifier?

burg replied:
Style: don't need -> ()

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -472,9 +473,15 @@ void WebAutomationSession::createBrowsingContext(std::optional<Inspector::Protoc
-        callback({ { protectedThis->handleForWebPageProxy(*page), toProtocol(protectedThis->m_client->currentPresentationOfPage(protectedThis.get(), *page)) } });
+        protectedPage->launchInitialProcessIfNecessary();
+
+#if ENABLE(WEBDRIVER_BIDI)
+        Ref process = protectedPage->legacyMainFrameProcess();
+        process->send(Messages::WebAutomationSessionProxy::EnsureRealmForInitialEmptyDocument(protectedPage->webPageIDInMainFrameProcess()), 0);
```
If this just exists to make the spec tests pass, It might be easier to synthesize this realm from UIProcess-side, so that the IPC send mid-create can be avoided.

sergedeh replied:
I implemented the solution by forcing eager creation of the JSWindowProxy to comply with the BiDi spec's realm lifecycle requirements. 

> The [remote end event trigger](https://w3c.github.io/webdriver-bidi/#event-remote-end-event-trigger) is:
When any of the [set up a window environment settings object](https://html.spec.whatwg.org/multipage/nav-history-apis.html#set-up-a-window-environment-settings-object), [set up a worker environment settings object](https://html.spec.whatwg.org/multipage/workers.html#set-up-a-worker-environment-settings-object) or [set up a worklet environment settings object](https://html.spec.whatwg.org/multipage/worklets.html#set-up-a-worklet-environment-settings-object) algorithms are invoked, immediately prior to returning the settings object:

While this approach follows the spec flow correctly, the ideal solution would preserve WebKit's lazy JSGlobalObject creation architecture (which I'm taking as an invariant).
I've added a FIXME comment linking to bug #310506 for implementing a better approach when we tackle more sophisticated realm types (workers/worklets). That future implementation should use for example a logical realm registry that tracks realm identifiers separately from JSGlobalObject materialization, allowing us to maintain both spec compliance and WebKit's lazy semantics.

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8227,6 +8225,53 @@ void WebPageProxy::didFinishLoadForFrame(IPC::Connection& connection, FrameIdent
 }
 
+#if ENABLE(WEBDRIVER_BIDI)
+void WebPageProxy::scriptRealmWasCreated(WebCore::FrameIdentifier frameID, String&& origin, CompletionHandler<void()>&& completionHandler)
```
These two methods should really be in the IPC between WebAutomationSession and its proxy in WebProcess. We shouldn't have WebPageProxy methods digging into the processor's agents and setting state, that should be handled via AutomationInstrumentation and its client which forwards from inside the session.

burg reviewed (changes requested):
Good work! Some polish needed, but this is heading in the right direction.

Regarding the dummy evaluation, it was OK to get the PR started, but I think we should find a way to avoid committing that.

Claude suggests:

> A more robust approach might be to directly access `ScriptController::windowProxy(mainWorld)` which forces proxy creation without executing script. Check if `coreFrame->script().windowProxy(mainThreadNormalWorld())` would trigger the same `createJSWindowProxy` path.

My suggestion inline was to just fake this identifier, but the suggestion above may suffice.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
```diff
@@ -349,6 +352,10 @@ friend class WebDriverBidiProcessor;
     // Called by WebAutomationSession messages.
     void logEntryAdded(const JSC::MessageSource&, const JSC::MessageLevel&, const String& messageText, const JSC::MessageType&, const WallTime&);
+#if ENABLE(WEBDRIVER_BIDI)
+    void scriptRealmCreated(WebCore::FrameIdentifier, uint64_t realmIdentifierValue, const WebCore::SecurityOriginData&);
```
This should be RealmIdentifier object id type, you'll need to add an entry to a .serialization.in file so the IPC generator knows about that type.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -1146,6 +1149,58 @@ void WebAutomationSessionProxy::addMessageToConsole(const JSC::MessageSource& so
+    auto realmIdentifier = RealmIdentifier::generate();
+    m_frameToRealmIdentifier.set(frameID, realmIdentifier);
+
+    protect(WebProcess::singleton().parentProcessConnection())->send(Messages::WebAutomationSession::ScriptRealmCreated(frameID, realmIdentifier.toUInt64(), origin), 0);
```
With proper IPC support we shouldn't need `.toUInt64()` here.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -1146,6 +1149,58 @@ void WebAutomationSessionProxy::addMessageToConsole(const JSC::MessageSource& so
+    // The better long-term solution is a logical realm registry that binds to JSGlobalObject when it
+    // naturally materializes, preserving both spec compliance and WebKit's lazy semantics. This hybrid
+    // approach should be prioritized for worker/worklet realm implementation where eager materialization
+    // poses greater risks. See https://bugs.webkit.org/show_bug.cgi?id=310506
```
Good.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -1146,6 +1149,58 @@ void WebAutomationSessionProxy::addMessageToConsole(const JSC::MessageSource& so
+    auto realmIdentifier = it->value;
+    m_frameToRealmIdentifier.remove(it);
+
+    protect(WebProcess::singleton().parentProcessConnection())->send(Messages::WebAutomationSession::ScriptRealmDestroyed(frameID, realmIdentifier.toUInt64()), 0);
```
Ditto.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
```diff
@@ -111,13 +112,19 @@ class WebAutomationSessionProxy : public IPC::MessageReceiver
     HashMap<WebCore::FrameIdentifier, HashMap<JSCallbackIdentifier, CompletionHandler<void(String&&, String&&)>>> m_webFramePendingEvaluateJavaScriptCallbacksMap;
     HashMap<WebCore::FrameIdentifier, Ref<WebAutomationDOMWindowObserver>> m_frameObservers;
+#if ENABLE(WEBDRIVER_BIDI)
+    HashMap<WebCore::FrameIdentifier, RealmIdentifier> m_frameToRealmIdentifier;
```
Nice.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -448,94 +435,52 @@ RefPtr<Inspector::Protocol::BidiScript::RealmInfo> BidiScriptAgent::createRealmI
+    if (it == m_browsingContextToRealmId.end()) {
+        // No realm has been created yet for this context - generate a new identifier.
+        // Note: This should not normally happen since notifyRealmCreated should be called
+        // before getRealms due to the sendWithAsyncReply barrier.
```
Didn't you remove sendWithAsyncReply for this?

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2975,6 +2993,34 @@ void WebAutomationSession::logEntryAdded(const JSC::MessageSource& messageSource
+void WebAutomationSession::scriptRealmDestroyed(WebCore::FrameIdentifier frameID, uint64_t realmIdentifierValue)
+{
+    RefPtr frame = WebFrameProxy::webFrame(frameID);
+    if (!frame)
```
Hmm..

`WindowProxy::detachFromFrame()` sets `m_frame = nullptr` before the loop that calls `destroyJSWindowProxy()`. The new code checks `if (m_frame)` before firing notifications -- but `m_frame` is already null. Every realm destruction via `detachFromFrame` is silently dropped.

```cpp
void WindowProxy::detachFromFrame()
{
    ASSERT(m_frame);

    m_frame = nullptr; // BJB> uh oh..

    // It's likely that destroying proxies will create a lot of garbage.
    if (!m_jsWindowProxies.isEmpty()) {
        do {
            auto it = m_jsWindowProxies.begin();
            it->value->window()->setConsoleClient(nullptr);
            destroyJSWindowProxy(it->key);
        } while (!m_jsWindowProxies.isEmpty());
        collectGarbageAfterWindowProxyDestruction();
    }
}
```

This is the primary destruction path for: tab close, cross-process navigation, iframe removal. The `contextDestroyedForPage` fallback covers main frames only. frame realm destruction during detach is missed entirely. `m_activeRealms` accumulates stale entries that get replayed on subscribe.

 Can you save away the `Frame` ref before it is cleared, or find another way to fix this?

burg reviewed (changes requested):
PR looks good overall, but as written the realmDestroyed notifications might not get sent for iframes and certain types of top-level navigations.

Since this PR is getting big, we may need to split iframes into a separate PR. I'm pretty sure we can get more iframe tests passing by fixing this.

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -3916,10 +3916,92 @@
+                "expected": { "all": { "status": ["FAIL"], "bug": "webkit.org/b/288067"}}
+            },
+            "test_subscribe_to_one_context": {
+                "expected": { "all": { "status": ["FAIL"], "bug": "webkit.org/b/282981"}}
```
What blocks this from progressing, since this PR implements event filtering? Is it blocked on the Event subscriptions PR?

sergedeh replied:
It was actually passing after the latest update. I have removed the line.

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -3916,10 +3916,92 @@
+            "test_shared_worker": {
+                "expected": { "all": { "status": ["FAIL"], "bug": "webkit.org/b/288067"}}
+            },
+            "test_service_worker": {
```
This test doesn't exist.

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -3916,10 +3916,92 @@
+            "test_reload": {
+                "expected": { "all": { "status": ["PASS"]}}
+            },
+            "test_existing_realm[window]": {
```
Do these need a bug link?

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -3916,10 +3916,92 @@
+            }
+        }
     },
     "imported/w3c/webdriver/tests/bidi/script/realm_destroyed/realm_destroyed.py": {
```
We don't support the sandbox related features, so should `test_subscribe_after_sandbox_creation` be marked FAIL?

sergedeh committed:
[WebDriver BiDi] Implement script.realmCreated and script.realmDestroyed events
https://bugs.webkit.org/show_bug.cgi?id=304062
rdar://166875337

Reviewed by BJ Burg.

This patch implements the script.realmCreated and script.realmDestroyed events
as specified in the W3C WebDriver BiDi specification. These events notify BiDi
clients when JavaScript execution realms are created and destroyed in browsing
contexts, enabling proper tracking of realm lifecycle for script execution.

* Source/WebCore/automation/AutomationInstrumentation.cpp:
(WebCore::AutomationInstrumentation::scriptRealmCreated): Added with world filtering.
(WebCore::AutomationInstrumentation::scriptRealmDestroyed): Added with world filtering.

* Source/WebCore/automation/AutomationInstrumentation.h:
Added scriptRealmCreated and scriptRealmDestroyed methods with DOMWrapperWorld parameter.

* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::resolveOriginForRealm): Added helper to resolve realm origin.
(WebCore::WindowProxy::destroyJSWindowProxy): Notify realm destruction, with optional frameForNotification parameter.
(WebCore::WindowProxy::createJSWindowProxy): Notify realm creation.
(WebCore::WindowProxy::setDOMWindow): Notify realm destruction and re-creation during navigation.
(WebCore::WindowProxy::detachFromFrame): Save frame reference before nullifying for realm destruction notifications.

* Source/WebCore/bindings/js/WindowProxy.h:
Updated destroyJSWindowProxy signature to accept optional Frame pointer for notifications during detachment.

* Source/WebDriver/Session.cpp:
(WebDriver::extractEventBrowsingContextIDs): Helper to extract context IDs from event parameters.
(WebDriver::haveSameElements): Helper to compare vectors for subscription matching.
(WebDriver::subscriptionMatchesScope): Check if subscription matches event scope.
(WebDriver::shouldReplayRealmCreatedEvents): Check if realm replay is needed.
(WebDriver::removeSubscribedEvents): Helper to decrement event subscription counts.
(WebDriver::Session::dispatchBidiMessage): Add context-based event filtering and realm event replay.
(WebDriver::Session::processLogEntryAdded): Extract event parameters for context filtering.

* Source/WebDriver/Session.h:
Changed emitActiveBidiScriptRealmCreatedEvents to return void.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::emitActiveBidiScriptRealmCreatedEvents): Emit realm events for active realms.
Handle subscribe command with realm event replay support.

* Source/WebKit/Scripts/webkit/messages.py:
Added WebKit::RealmIdentifier to serialized_identifiers() list for proper IPC message generation.

* Source/WebKit/Shared/IdentifierTypes.h:
Added RealmIdentifier as ObjectIdentifier<RealmIdentifierType>.

* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
Added RealmIdentifierType template for ObjectIdentifier serialization.

* Source/WebKit/UIProcess/Automation/Automation.json:
Added emitActiveBidiScriptRealmCreatedEvents command.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::notifyRealmCreated): Track realm and emit script.realmCreated event.
(WebKit::BidiScriptAgent::notifyRealmDestroyed): Remove realm and emit script.realmDestroyed event.
(WebKit::BidiScriptAgent::realmIdentifierForBrowsingContext): Lookup helper for UIProcess-initiated destruction.
(WebKit::BidiScriptAgent::generateRealmIdForBrowsingContext): Generate realm ID string from identifier.
(WebKit::BidiScriptAgent::emitEventsForActiveRealms): Emit realm events for all active realms.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
Added RealmInfo struct for tracking realm metadata.
Added RealmIdentifier-based realm tracking (m_browsingContextToRealmId, m_activeRealms).
Added notifyRealmCreated, notifyRealmDestroyed, realmIdentifierForBrowsingContext, and emitEventsForActiveRealms methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::createBrowsingContext): Call ensureRealmForInitialEmptyDocument.
(WebKit::WebAutomationSession::scriptRealmCreated): Handle realm creation IPC from WebProcess using RealmIdentifier.
(WebKit::WebAutomationSession::scriptRealmDestroyed): Handle realm destruction IPC with activeRealms lookup to avoid race conditions.
(WebKit::WebAutomationSession::contextDestroyedForPage): Lookup realm ID before notifying destruction.
(WebKit::WebAutomationSession::emitActiveBidiScriptRealmCreatedEvents): Delegate to BidiScriptAgent.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Added scriptRealmCreated, scriptRealmDestroyed, and emitActiveBidiScriptRealmCreatedEvents handler declarations using RealmIdentifier type.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.messages.in:
Added ScriptRealmCreated and ScriptRealmDestroyed IPC messages with RealmIdentifier parameter.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor): Initialize m_scriptDomainNotifier.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
Added scriptAgent() and scriptDomainNotifier() accessors.
Added m_scriptDomainNotifier member.

* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
Added realmCreated and realmDestroyed event definitions.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::scriptRealmCreated): Forward to WebAutomationSession via IPC with RealmIdentifier.
(WebKit::WebAutomationSessionProxy::scriptRealmDestroyed): Forward to WebAutomationSession via IPC with RealmIdentifier.
(WebKit::WebAutomationSessionProxy::ensureRealmForInitialEmptyDocument): Force JSWindowProxy creation using jsWindowProxy().

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
Added AutomationInstrumentationClient interface implementation.
Added ensureRealmForInitialEmptyDocument method.
Added m_frameToRealmIdentifier mapping.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
Added EnsureRealmForInitialEmptyDocument IPC message.

* WebDriverTests/TestExpectations.json:
Updated test expectations for realm_created tests to match current implementation.
Added missing test expectations for sandbox, workers, and iframe tests.

Canonical link: https://commits.webkit.org/310527@main
